services/scan/src/precinct_scanner_state_machine.ts

```diff
@@ -0,0 +1,854 @@
+        },
+        SCANNER_READY_TO_SCAN: 'returned',
+        SCANNER_READY_TO_EJECT: 'error_jammed',
+      },
+      // But, if you pull the paper out right after rejecting, we go straight to
+      // NO_PAPER, skipping READY_TO_SCAN completely. So we need to eventually
+      // timeout waiting for READY_TO_SCAN.
+      after: { 1000: 'no_paper' },
+    },
+    returned: {
+      invoke: pollPaperStatus,
+      on: {
+        SCANNER_READY_TO_SCAN: { target: 'returned', internal: true },
+        SCANNER_NO_PAPER: 'no_paper',
+      },
+    },
+    rejecting: {
+      invoke: {
+        src: reject,
+        onDone: 'checking_rejecting_completed',
+        onError: 'error_jammed',
+      },
+    },
+    // After rejecting, before the plustek grabs the paper to hold it, it sends
+    // NO_PAPER status for a bit before sending READY_TO_SCAN. So we need to
+    // wait for the READY_TO_SCAN.
+    checking_rejecting_completed: {
+      invoke: pollPaperStatus,
+      on: {
+        SCANNER_NO_PAPER: {
+          target: 'checking_rejecting_completed',
+          internal: true,
+        },
+        SCANNER_READY_TO_SCAN: 'rejected',
+        SCANNER_READY_TO_EJECT: 'error_jammed',
+      },
+      // But, if you pull the paper out right after rejecting, we go straight to
+      // NO_PAPER, skipping READY_TO_SCAN completely. So we need to eventually
+      // timeout waiting for READY_TO_SCAN.
+      after: { 1000: 'no_paper' },
+    },
+    // Paper has been rejected and is held in the front, waiting for removal.
+    rejected: {
+      invoke: pollPaperStatus,
+      on: {
+        SCANNER_READY_TO_SCAN: { target: 'rejected', internal: true },
+        SCANNER_NO_PAPER: 'no_paper',
+      },
+    },
+    calibrating: {
+      entry: assign({ error: undefined }),
+      invoke: {
+        src: calibrate,
+        onDone: 'calibrated',
+        onError: {
+          target: 'calibrated',
+          actions: assign({ error: (_context, event) => event.data }),
+        },
+      },
+    },
+    calibrated: {
+      on: {
+        START_OVER: 'no_paper',
+      },
+    },
+    error_jammed: {
+      invoke: pollPaperStatus,
+      on: {
+        SCANNER_NO_PAPER: 'no_paper',
+        SCANNER_JAM: { target: 'error_jammed', internal: true },
+        SCANNER_READY_TO_SCAN: { target: 'error_jammed', internal: true },
+        SCANNER_READY_TO_EJECT: { target: 'error_jammed', internal: true },
+      },
+    },
+    error_both_sides_have_paper: {
+      entry: assign({
+        error: new PrecinctScannerError('both_sides_have_paper'),
+      }),
+      invoke: pollPaperStatus,
+      on: {
+        SCANNER_BOTH_SIDES_HAVE_PAPER: {
+          target: 'error_both_sides_have_paper',
+          internal: true,
+        },
+        // For now, if the front paper is removed, just reject the back paper,
+        // since we don't have context on how we got here and what was supposed
+        // to happen.
+        SCANNER_READY_TO_EJECT: 'rejecting',
```

Comment:
This is probably overly conservative, as the main way this will happen is if you put in a second sheet right before the first sheet is accepted, so removing the second sheet from intake should then cause the first sheet to resume being accepted. But, we can optimize this later. This isn't wrong per se, just overly conservative.

Reply:
Yep, I agree completely. My goal on this first pass was to make sure we never ended up in an inconsistent state.

I think we should for sure come back and improve the user experience here. There is a fancy [history](https://xstate.js.org/docs/guides/history.html#history-state-configuration) feature in xstate that would allow this to be easily changed, but I was/am hesitant to bring in fancier state machine features if we don't have to.